Remove unused flags from kubectl run

brianpursley opened the pull request:
#### What type of PR is this?
/kind bug
/kind cleanup

#### What this PR does / why we need it:
The following flags, which do not apply to kubectl run, have been removed:
* --cascade
* --filename
* --force
* --grace-period
* --kustomize
* --recursive
* --timeout
* --wait

These flags were being added to the run command to support pod deletion after attach, but they are not used if set, so they effectively do nothing.

This PR also displays an error message if the pod fails to be deleted (when the --rm flag is used).  Previously any error during deletion would be suppressed and the pod would remain.

This PR also adds some unit tests for run and attach with and without the --rm flag.  As such, some minor refactoring of the run command has been done to support mocking dependencies.

#### Which issue(s) this PR fixes:
Fixes #108630

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
```release-note
Removed unused flags from kubectl run command
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:
```docs

```

brianpursley commented:
/retest

ardaguclu reviewed (commented):
That looks good to me but I'd expect removal of this line https://github.com/kubernetes/kubernetes/blob/202d1e6fc5993b93ef6100f0e3f516a249386f4a/staging/src/k8s.io/kubectl/pkg/cmd/run/run.go#L167. This PR still exposes delete flags?

brianpursley commented:
> That looks good to me but I'd expect removal of this line
> 
> https://github.com/kubernetes/kubernetes/blob/202d1e6fc5993b93ef6100f0e3f516a249386f4a/staging/src/k8s.io/kubectl/pkg/cmd/run/run.go#L167
> 
> . This PR still exposes delete flags?

Yeah.  [`DeleteFlags.AddFlags()`](https://github.com/kubernetes/kubernetes/blob/8f8138bc35f59ef04b77a3520d8d1065603ccdb7/staging/src/k8s.io/kubectl/pkg/cmd/delete/delete_flags.go#L112-L158) has some logic that excludes any flag that does not have a default set.

I was approaching this PR with a mindset of trying to keep things as similar to how they are currently as possible.

But after doing this, I wonder if it might be better to simply remove the DeleteFlags altogether.  I will circle back and take a look at what they would look like.

brianpursley commented:
By the way, here is the help output now, with the flags removed:
```
~ $ kubectl run --help
Create and run a particular image in a pod.

Examples:
  # Start a nginx pod
  kubectl run nginx --image=nginx
  
  # Start a hazelcast pod and let the container expose port 5701
  kubectl run hazelcast --image=hazelcast/hazelcast --port=5701
  
  # Start a hazelcast pod and set environment variables "DNS_DOMAIN=cluster" and "POD_NAMESPACE=default" in the
container
  kubectl run hazelcast --image=hazelcast/hazelcast --env="DNS_DOMAIN=cluster" --env="POD_NAMESPACE=default"
  
  # Start a hazelcast pod and set labels "app=hazelcast" and "env=prod" in the container
  kubectl run hazelcast --image=hazelcast/hazelcast --labels="app=hazelcast,env=prod"
  
  # Dry run; print the corresponding API objects without creating them
  kubectl run nginx --image=nginx --dry-run=client
  
  # Start a nginx pod, but overload the spec with a partial set of values parsed from JSON
  kubectl run nginx --image=nginx --overrides='{ "apiVersion": "v1", "spec": { ... } }'
  
  # Start a busybox pod and keep it in the foreground, don't restart it if it exits
  kubectl run -i -t busybox --image=busybox --restart=Never
  
  # Start the nginx pod using the default command, but use custom arguments (arg1 .. argN) for that command
  kubectl run nginx --image=nginx -- <arg1> <arg2> ... <argN>
  
  # Start the nginx pod using a different command and custom arguments
  kubectl run nginx --image=nginx --command -- <cmd> <arg1> ... <argN>

Options:
    --allow-missing-template-keys=true:
	If true, ignore any errors in templates when a field or map key is missing in the template. Only applies to
	golang and jsonpath output formats.

    --annotations=[]:
	Annotations to apply to the pod.

    --attach=false:
	If true, wait for the Pod to start running, and then attach to the Pod as if 'kubectl attach ...' were called.
	Default false, unless '-i/--stdin' is set, in which case the default is true. With '--restart=Never' the exit
	code of the container process is returned.

    --command=false:
	If true and extra arguments are present, use them as the 'command' field in the container, rather than the
	'args' field which is the default.

    --dry-run='none':
	Must be "none", "server", or "client". If client strategy, only print the object that would be sent, without
	sending it. If server strategy, submit server-side request without persisting the resource.

    --env=[]:
	Environment variables to set in the container.

    --expose=false:
	If true, create a ClusterIP service associated with the pod.  Requires `--port`.

    --field-manager='kubectl-run':
	Name of the manager used to track field ownership.

    --image='':
	The image for the container to run.

    --image-pull-policy='':
	The image pull policy for the container.  If left empty, this value will not be specified by the client and
	defaulted by the server.

    -l, --labels='':
	Comma separated labels to apply to the pod. Will override previous values.

    --leave-stdin-open=false:
	If the pod is started in interactive mode or with stdin, leave stdin open after the first attach completes. By
	default, stdin will be closed after the first attach completes.

    -o, --output='':
	Output format. One of: (json, yaml, name, go-template, go-template-file, template, templatefile, jsonpath,
	jsonpath-as-json, jsonpath-file).

    --override-type='merge':
	The method used to override the generated object: json, merge, or strategic.

    --overrides='':
	An inline JSON override for the generated object. If this is non-empty, it is used to override the generated
	object. Requires that the object supply a valid apiVersion field.

    --pod-running-timeout=1m0s:
	The length of time (like 5s, 2m, or 3h, higher than zero) to wait until at least one pod is running

    --port='':
	The port that this container exposes.

    --privileged=false:
	If true, run the container in privileged mode.

    -q, --quiet=false:
	If true, suppress prompt messages.

    --restart='Always':
	The restart policy for this Pod.  Legal values [Always, OnFailure, Never].

    --rm=false:
	If true, delete the pod after it exits.  Only valid when attaching to the container, e.g. with '--attach' or
	with '-i/--stdin'.

    --save-config=false:
	If true, the configuration of current object will be saved in its annotation. Otherwise, the annotation will
	be unchanged. This flag is useful when you want to perform kubectl apply on this object in the future.

    --show-managed-fields=false:
	If true, keep the managedFields when printing objects in JSON or YAML format.

    -i, --stdin=false:
	Keep stdin open on the container in the pod, even if nothing is attached.

    --template='':
	Template string or path to template file to use when -o=go-template, -o=go-template-file. The template format
	is golang templates [http://golang.org/pkg/text/template/#pkg-overview].

    --timeout=0s:
	The length of time to wait before giving up on a delete, zero means determine a timeout from the size of the
	object

    -t, --tty=false:
	Allocate a TTY for the container in the pod.

Usage:
  kubectl run NAME --image=image [--env="key=value"] [--port=port] [--dry-run=server|client] [--overrides=inline-json]
[--command] -- [COMMAND] [args...] [options]

Use "kubectl options" for a list of global command-line options (applies to all commands).
```

brianpursley commented:
@ardaguclu I removed DeleteFlags altogether.  I think it simplifies things even more.  Thanks for the suggestion.  PTAL.

ardaguclu commented:
> @ardaguclu I removed DeleteFlags altogether. I think it simplifies things even more. Thanks for the suggestion. PTAL.

That looks good to me. I'll label it later for the other reviewers have a chance to look at.

ardaguclu commented on staging/src/k8s.io/kubectl/pkg/cmd/run/run_test.go:
```diff
@@ -20,6 +20,9 @@ import (
 	"io/ioutil"
+	schema "k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/tools/remotecommand"
+	"k8s.io/kubectl/pkg/cmd/attach"
```
I have exact same problem :)

brianpursley committed:
Remove unused flags from kubectl run

The following flags, which do not apply to kubectl run,
have been removed:
--cascade
--filename
--force
--grace-period
--kustomize
--recursive
--timeout
--wait

These flags were being added to the run command to support
pod deletion after attach, but they are not used if set, so
they effectively do nothing.

This PR also displays an error message if the pod fails to be
deleted (when the --rm flag is used).  Previously any error
during deletion would be suppressed and the pod would remain.

This PR also adds some unit tests for run and attach with and
without the --rm flag.  As such, some minor refactoring of the
run command has been done to support mocking dependencies.

brianpursley commented:
/retest

ardaguclu commented:
/triage accepted

ardaguclu commented:
/lgtm

saschagrunert commented:
Hey folks, we come across this during our k8s upgrade to v1.25 :blush: I think having the information that the flags are unused and which have being removed would be great to capture in the release notes.

brianpursley commented:
> Hey folks, we come across this during our k8s upgrade to v1.25 blush I think having the information that the flags are unused and which have being removed would be great to capture in the release notes.

Actually, I think it was a mistake to completely remove the flags at this time, and this PR needs to be reverted.

While the flags are completely useless, they do exist, and if someone happened to be setting them, their `kubectl run` command would still work.  However starting in 1.25, the command will now fail with an `unknown flag` error.

Even though the change is a good one to make, it hasn't gone through the required deprecation period, and could break backward compatibility if someone happened to be using those flags.

ardaguclu commented:
> > Hey folks, we come across this during our k8s upgrade to v1.25 blush I think having the information that the flags are unused and which have being removed would be great to capture in the release notes.
> 
> Actually, I think it was a mistake to completely remove the flags at this time, and this PR needs to be reverted.
> 
> While the flags are completely useless, they do exist, and if someone happened to be setting them, their `kubectl run` command would still work. However starting in 1.25, the command will now fail with an `unknown flag` error.
> 
> Even though the change is a good one to make, it hasn't gone through the required deprecation period, and could break backward compatibility if someone happened to be using those flags.

Wouldn't be better adding manual deprecation for flags in run command instead directly reverting this useful fix?. These commands are redundant and we just don't want to break any script. In any case, there will be a backport to 1.25.

brianpursley commented:
> > > Hey folks, we come across this during our k8s upgrade to v1.25 blush I think having the information that the flags are unused and which have being removed would be great to capture in the release notes.
> > 
> > 
> > Actually, I think it was a mistake to completely remove the flags at this time, and this PR needs to be reverted.
> > While the flags are completely useless, they do exist, and if someone happened to be setting them, their `kubectl run` command would still work. However starting in 1.25, the command will now fail with an `unknown flag` error.
> > Even though the change is a good one to make, it hasn't gone through the required deprecation period, and could break backward compatibility if someone happened to be using those flags.
> 
> Wouldn't be better adding manual deprecation for flags in run command instead directly reverting this useful fix?. These commands are redundant and we just don't want to break any script. In any case, there will be a backport to 1.25.

Ideally, but I think process-wise, it makes sense to revert this and then do the deprecation in a new PR.

This will allow time for the re-do PR to be reviewed.
